test/e2e/k8s_util.go

```diff
@@ -214,11 +214,80 @@ func (k *KubernetesUtils) CreateOrUpdateDeployment(ns, deploymentName string, re
 	return d, err
 }
 
+// BuildService is a convenience function for building a corev1.Service spec.
+func (k *KubernetesUtils) BuildService(svcName, svcNS string, port, targetPort int, selector map[string]string, serviceType *v1.ServiceType) *v1.Service {
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      svcName,
+			Namespace: svcNS,
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Port:       int32(port),
+				TargetPort: intstr.FromInt(targetPort),
+			}},
+			Selector: selector,
+		},
+	}
+	if serviceType != nil {
+		service.Spec.Type = *serviceType
+	}
+	return service
+}
+
+// CreateOrUpdateService is a convenience function for updating/creating Services.
+func (k *KubernetesUtils) CreateOrUpdateService(svc *v1.Service) (*v1.Service, error) {
+	log.Infof("creating/updating Service %s in ns %s", svc.Name, svc.Namespace)
+	svcReturned, err := k.clientset.CoreV1().Services(svc.Namespace).Get(context.TODO(), svc.Name, metav1.GetOptions{})
```

Comment:
I noticed that for other similar functions (e.g. `CreateOrUpdateNamespace`), we call `Create` directly and then call `Update` if the creation fails. We don't bother with calling `Get` first to check if the resource already exists.

Reply:
For Service we cannot use our built Service object to update it: `[metadata.resourceVersion: Invalid value: "": must be specified for an update, spec.clusterIPs[0]: Invalid value: []string(nil): primary clusterIP can not be unset]` This is not the case for resources like Namespace.